server: new statement details endpoint

maryliag opened the pull request:
New endpoint for Statement Details. The endpoint returns:
- Statement: the total stats for that statement fingerprint
- StatementsPerAggregatedTs: returns the statement from the period
selected but each aggregated timestamp is an entry. This list
can be used on the overview page on the Statement Details for
the creation of charts by execution timestamp.
- StatementsPerPlanHash: returns the statement from the period
selected but each plan hash is an entry. This list can be used
on the Explain Plan tab of the Statement Details, to show the
execution stats per Plan.

Partially addresses #72129

Release note (api change): New Statement Details endpoint that returns
the details for a selected statement, and its execution separated by
aggregation timestamp (statementsPerAggregatedTs) and by plan hash
(statementsPerPlanHash).

Release Justification: Low risk, high benefit change

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76592)
<!-- Reviewable:end -->

maryliag committed:
server: new statement details endpoint

New endpoint for Statement Details. The endpoint returns:
- Statement: the total stats for that statement fingerprint
- StatementsPerAggregatedTs: returns the statement from the period
selected but each aggregated timestamp is an entry. This list
can be used on the overview page on the Statement Details for
the creation of charts by aggregation timestamp.
- StatementsPerPlanHash: returns the statement from the period
selected but each plan hash is an entry. This list can be used
on the Explain Plan tab of the Statement Details, to show the
execution stats per Plan.

Partially addresses #72129

Release note (api change): New Statement Details endpoint that returns
the details for a selected statement, and its execution separated by
aggregation timestamp (statementsPerAggregatedTs) and by plan hash
(statementsPerPlanHash).

Azhng reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76592)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng and @maryliag)

___
*[-- commits, line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwXJn3O58QVTSZnAZcs):*
Nit: we don't really have the concept of `execution timestamp` since that information is already lost by the time we collect those stmt stats. Do you mean `aggregation timestamp` ?
___
*[pkg/server/serverpb/status.proto, line 1326 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwXHa_ODZWRpyNasnhJ:-MwXHa_ODZWRpyNasnhK:b-mr64ou) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56b9796795d8a3171cabb821140ca965698a102/pkg/server/serverpb/status.proto#L1326)):*
> ```protobuf
>   // query, failed, implicitTxn and database. So we don't need to add them
>   // to the request.
>   uint64 fingerprint_id = 1;
> ```

Should we also add in the transaction fingerprint ID here? Since the same statement stats can be part of multiple different transactions 
___
*[pkg/server/serverpb/status.proto, line 1330 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwXIBJC8C4gAgmbY4m_:-MwXIBJC8C4gAgmbY4ma:bvgmq1g) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56b9796795d8a3171cabb821140ca965698a102/pkg/server/serverpb/status.proto#L1330)):*
> ```protobuf
>   string full_scan = 3;
>   string distSQL = 4;
>   string vec = 5;
> ```

`full_scan`, `distSQL` and `vec` are essentially the details information of a query plan, which is encoded in the `plan_hash`. I feel that since we are already loading multiple plans for the statement details page, perhaps we don't really need these three fields? 
___
*[pkg/server/serverpb/status.proto, line 1333 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwXHxCQ7HnWvZrEQE1q:-MwXHxCQ7HnWvZrEQE1r:b-lc0gto) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56b9796795d8a3171cabb821140ca965698a102/pkg/server/serverpb/status.proto#L1333)):*
> ```protobuf
>   // Unix time range for aggregated statements.
>   int64 start = 6 [(gogoproto.nullable) = true];
>   int64 end = 7 [(gogoproto.nullable) = true];
> ```

Most of the protobuf in CRDB define the "time" type using `google.protobuf.Timestamp` type with some `gogoproto` override. Perhaps we can use that here instead of `int64`. 

E.g.

```proto
google.protobuf.Timestamp created_at = 5 [(gogoproto.stdtime) = true];
```
___
*[pkg/server/serverpb/status.proto, line 1350 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwXJOWF8uAHq40mCzyf:-MwXJOWF8uAHq40mCzyg:b-3sdral) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56b9796795d8a3171cabb821140ca965698a102/pkg/server/serverpb/status.proto#L1350)):*
> ```protobuf
>       (gogoproto.stdduration) = true];
>     google.protobuf.Timestamp aggregated_ts = 3 [(gogoproto.nullable) = false, (gogoproto.stdtime) = true];
>     string explain_plan = 4;
> ```

Hmm is this field meant to be used to store the sampled plans? I think historically we have been using the `ExplainTreePlanNode` for this field. 
___
*[pkg/server/serverpb/status.proto, line 1351 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwXKT-07R9AZuaqjqSX:-MwXKT-07R9AZuaqjqSY:b8uhhn6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56b9796795d8a3171cabb821140ca965698a102/pkg/server/serverpb/status.proto#L1351)):*
> ```protobuf
>     google.protobuf.Timestamp aggregated_ts = 3 [(gogoproto.nullable) = false, (gogoproto.stdtime) = true];
>     string explain_plan = 4;
>     uint64 plan_hash = 5;
> ```

also might worth a comment here noting what it means when this field is null. 
___
*[pkg/server/serverpb/status.proto, line 1358 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwXK8sSB0aC1jPLkHTq:-MwXK8sSB0aC1jPLkHTr:b-wmi3s6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56b9796795d8a3171cabb821140ca965698a102/pkg/server/serverpb/status.proto#L1358)):*
> ```protobuf
>   CollectedStatement statement = 1 [(gogoproto.nullable) = false];
>   repeated CollectedStatementGrouped statements_per_aggregated_ts = 2 [(gogoproto.nullable) = false];
>   repeated CollectedStatementGrouped statements_per_plan_hash = 3 [(gogoproto.nullable) = false];
> ```

Hmm is this field populated for "every distinct plan hash" or "every distinct plan hash and every distinct aggregated_ts" ? Might worth a comment here. 
___
*[pkg/server/serverpb/status.proto, line 1361 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwXIteXBIWSB8hOf77r:-MwXIteXBIWSB8hOf77s:bb8z1yq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56b9796795d8a3171cabb821140ca965698a102/pkg/server/serverpb/status.proto#L1361)):*
> ```protobuf
>   // If set and non-empty, indicates the prefix to application_name
>   // used for statements/queries issued internally by CockroachDB.
>   string internal_app_name_prefix = 4;
> ```

hmm do we want to allow this to be `nullable` ?


<!-- Sent from Reviewable.io -->

maryliag reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76592)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng)

___
*[-- commits, line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwXJn3O58QVTSZnAZcs):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Nit: we don't really have the concept of `execution timestamp` since that information is already lost by the time we collect those stmt stats. Do you mean `aggregation timestamp` ?
</blockquote></details>

Yes, I meant the aggregation. Made the changes
___
*[pkg/server/serverpb/status.proto, line 1326 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwXHa_ODZWRpyNasnhJ:-MwY1ba_-lT0BMNd8u0_:b5lm0xu) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56b9796795d8a3171cabb821140ca965698a102/pkg/server/serverpb/status.proto#L1326)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Should we also add in the transaction fingerprint ID here? Since the same statement stats can be part of multiple different transactions 
</blockquote></details>

On the statement page we don't care about the txn fingerprint, only on the transaction details page is that we split
___
*[pkg/server/serverpb/status.proto, line 1330 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwXIBJC8C4gAgmbY4m_:-MwY1mEH25jKhOTiQ8IQ:b5cr1d4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56b9796795d8a3171cabb821140ca965698a102/pkg/server/serverpb/status.proto#L1330)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

`full_scan`, `distSQL` and `vec` are essentially the details information of a query plan, which is encoded in the `plan_hash`. I feel that since we are already loading multiple plans for the statement details page, perhaps we don't really need these three fields? 
</blockquote></details>

oh I was not aware those were included on plan_hash, so that makes it easier and I can remove it from here, np! 
___
*[pkg/server/serverpb/status.proto, line 1333 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwXHxCQ7HnWvZrEQE1q:-MwY1zQW2cVq4Bgb8lmv:bcxabua) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56b9796795d8a3171cabb821140ca965698a102/pkg/server/serverpb/status.proto#L1333)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Most of the protobuf in CRDB define the "time" type using `google.protobuf.Timestamp` type with some `gogoproto` override. Perhaps we can use that here instead of `int64`. 

E.g.

```proto
google.protobuf.Timestamp created_at = 5 [(gogoproto.stdtime) = true];
```
</blockquote></details>

I'm following the value used on the combinedstats endpoint, because what I'm doing is getting the same value from there to use here. If we keep each one with one format it would be messy to align. We can reconsider on a future PR, but I'll keep this one as is.
___
*[pkg/server/serverpb/status.proto, line 1350 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwXJOWF8uAHq40mCzyf:-MwY5sRL5wQhZGvHm1vU:bqesid6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56b9796795d8a3171cabb821140ca965698a102/pkg/server/serverpb/status.proto#L1350)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Hmm is this field meant to be used to store the sampled plans? I think historically we have been using the `ExplainTreePlanNode` for this field. 
</blockquote></details>

That is correct for plans saved on our stats, where you select the plan and it's returned in one line as a json that you can decode to tree, but this one here comes form the `crdb_internal.decode_plan_gist()`, which returns a row per line, with them being strings, they're not rendered
___
*[pkg/server/serverpb/status.proto, line 1351 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwXKT-07R9AZuaqjqSX:-MwYJkCH0G4Uq43XMTxi:bpfj2y8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56b9796795d8a3171cabb821140ca965698a102/pkg/server/serverpb/status.proto#L1351)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

also might worth a comment here noting what it means when this field is null. 
</blockquote></details>

I split into two formats here, to make it more clear
___
*[pkg/server/serverpb/status.proto, line 1358 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwXK8sSB0aC1jPLkHTq:-MwYJgzS1prNah3hJ58y:b-ya4zny) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56b9796795d8a3171cabb821140ca965698a102/pkg/server/serverpb/status.proto#L1358)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Hmm is this field populated for "every distinct plan hash" or "every distinct plan hash and every distinct aggregated_ts" ? Might worth a comment here. 
</blockquote></details>

Improved the comments
___
*[pkg/server/serverpb/status.proto, line 1361 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwXIteXBIWSB8hOf77r:-MwYJ_su86wZvUNdFmI6:b6jebjt) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56b9796795d8a3171cabb821140ca965698a102/pkg/server/serverpb/status.proto#L1361)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

hmm do we want to allow this to be `nullable` ?
</blockquote></details>

It can be nullable, which is also the pattern for all other internal_app_name_prefix usages in proto files.


<!-- Sent from Reviewable.io -->

jocrl reviewed (commented):
Thank you! I'll slack a separate question about transaction details 🙂.

I don't see an existing `combined_statement_stats_test.go`; do we not usually test endpoints (perhaps those that are simple enough)? Or did we just happen to not test the combined statements endpoint? If it's the latter, I think the new details endpoint should be tested, whether now or soon after the release.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-:-MwbCnQB1zA9B0ZPRSF-:b-2visym)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng)



<!-- Sent from Reviewable.io -->

maryliag reviewed (commented):
The frontend code has some parts to make testing endpoint easier, which is where we have the tests for the combined statements, so I added tests for the endpoint on my next PR

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-:-Mwg3lmt63xeXzIVYmhE:b-egy92o)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng)



<!-- Sent from Reviewable.io -->

matthewtodd reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76592)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @maryliag, and @matthewtodd)

___
*[pkg/server/combined_statement_stats.go, line 351 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwgfauW5GFvmirKwZii:-MwgfauW5GFvmirKwZij:b-7ywrh1) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L351)):*
> ```Go
> }
> 
> func getStatementDetailsQueryClausesAndArgs(
> ```

nit: maybe move this function below `getStatementDetails` for reading flow?
___
*[pkg/server/combined_statement_stats.go, line 357 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mwgf2Ty-419F30fpqil:-Mwgf2Ty-419F30fpqim:bbp8plk) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L357)):*
> ```Go
> 	buffer.WriteString(testingKnobs.GetAOSTClause())
> 
> 	buffer.WriteString(fmt.Sprintf(" WHERE encode(fingerprint_id, 'hex') = $%d", len(args)+1))
> ```

Can you say more about the `len(args)+1` here and below? Is this code building a query "template" with embedded `$1` placeholders that will be filled in at execution time? If so, it's not clear to me why the same value is used throughout. EDIT: I see now that this function is also modifying `args` as it goes along, and that behavior surprised me. I hope there's a plainer way to write the code (I haven't taken a moment to think yet), but if not it's worth a comment highlighting the mechanism.
___
*[pkg/server/combined_statement_stats.go, line 372 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mwgg-C7E38eITFwszig:-Mwgg-C7E38eITFwszih:b-uhgjrd) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L372)):*
> ```Go
> 		appNames := strings.Split(req.AppName, ",")
> 		if strings.Contains(req.AppName, "(unset)") {
> 			appNames = append(appNames, "")
> ```

Does this mean we'll pass along both `(unset)` and `` (the empty string)? That feels suspicious to me.
___
*[pkg/server/combined_statement_stats.go, line 376 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwggOny6joHdaszeR3a:-MwggOny6joHdaszeR3b:b-dqzz9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L376)):*
> ```Go
> 		buffer.WriteString(" AND (")
> 		for i, app := range appNames {
> 			if i != 0 {
> ```

Can this be done with a `strings.Join` or something? Except the modifying args in-place prevents that. Hmm.


<!-- Sent from Reviewable.io -->

Azhng reviewed (changes requested):
Nice. Do you know how long does it take to fetch this endpoint on the cluster with a lot of data?

> Reviewed 2 of 5 files at r1, 2 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-:-Mwh2bALDDJyFQ4PJ18p:b-cnvb3s)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @maryliag, and @matthewtodd)

___
*[pkg/server/combined_statement_stats.go, line 384 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mwh-B-c6GWXClFWPYOv:-Mwh-B-c6GWXClFWPYOw:b9l54jl) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L384)):*
> ```Go
> 			}
> 		}
> 		buffer.WriteString(" )")
> ```

How about using the `IN` operator here. 
```
WHERE app_name IN (<val1>, <val2>, <val3>)
```
___
*[pkg/server/combined_statement_stats.go, line 459 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mwh03Ia6f_8jkRIaLNP:-Mwh03Ia6f_8jkRIaLNQ:b-4g6s5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L459)):*
> ```Go
> 	var statement serverpb.StatementDetailsResponse_CollectedStatement
> 
> 	it, err := ie.QueryIteratorEx(ctx, "combined-stmts-details-total", nil,
> ```

If we are only querying for one row, then we can use `QueryRowEx` here instead.
___
*[pkg/server/combined_statement_stats.go, line 479 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mwh-tf14Kle4e9qIFJQ:-Mwh-tf14Kle4e9qIFJR:bjazs1v) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L479)):*
> ```Go
> 		var row tree.Datums
> 		if row = it.Cur(); row == nil {
> 			return statement, errors.New("unexpected null row")
> ```

probably should be returning `serverError` here and below too as well
___
*[pkg/server/combined_statement_stats.go, line 525 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mwh1DdqCwyjAF1uANiP:-Mwh1DdqCwyjAF1uANiQ:bl2xt7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L525)):*
> ```Go
> 			},
> 			AppNames:            appNames,
> 			AggregationInterval: time.Duration(aggInterval.Nanos()),
> ```

Simply assigning the `aggInterval` directly here
___
*[pkg/server/combined_statement_stats.go, line 527 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mwh0dwl3vvg1Lk3_FCB:-Mwh0dwl3vvg1Lk3_FCC:bloce0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L527)):*
> ```Go
> 			AggregationInterval: time.Duration(aggInterval.Nanos()),
> 			Stats:               metadata.Stats,
> 		}
> ```

nit: override the `metadata.Key.Query = queryPrettify` first, then you can simplify the assignment to be 

```golang

		statement = serverpb.StatementDetailsResponse_CollectedStatement{
			KeyData:             metadata.Key,
			AppNames:            appNames,
			AggregationInterval: time.Duration(aggInterval.Nanos()),
			Stats:               metadata.Stats,
		}

``` 


also, the `metadata` object here in fact also contains stats, might as well just call it `stats`. (I know we have been calling it `metadata` in other places of the code base, but that's really a mistake) 
___
*[pkg/server/combined_statement_stats.go, line 632 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mwh1yDW4Vlms1Ji8ExU:-Mwh1yDW4Vlms1Ji8ExV:b-9fxb2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L632)):*
> ```Go
> func getExplainPlanFromGist(ctx context.Context, ie *sql.InternalExecutor, planGist string) string {
> 	planError := "Error collecting Explain Plan."
> 	var args []interface{}
> ```

no need to use a slice here, since we know we only have 1 arg. 
___
*[pkg/server/serverpb/status.proto, line 1326 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mwgz1cV66ULWpiwb_25:-Mwgz1cV66ULWpiwb_26:b-cez416) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/serverpb/status.proto#L1326)):*
> ```protobuf
>   // query, failed, implicitTxn and database. So we don't need to add them
>   // to the request.
>   string fingerprint_id = 1;
> ```

nit: use the `gogoproto.customname` to override the Golang's field name to be `FingerprintID`, otherwise protoc will generate field name as `FingerprintId` 
___
*[pkg/server/serverpb/status.proto, line 1327 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwgzW5K-o-2yCCWRc3a:-MwgzW5K-o-2yCCWRc3b:bkei4oi) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/serverpb/status.proto#L1327)):*
> ```protobuf
>   // to the request.
>   string fingerprint_id = 1;
>   string app_name = 2;
> ```

Just realized that this is not a single `app_name`, but a list of `app_name`s sent from the frontend. My would much prefer to have `repeated string app_names = 2;` here. 

String manipulation and the special logic to deal with the `(unset)` case is exclusively frontend concept. Leaking frontend concept into the backend code is only going to make the code harder to maintain going forward.
___
*[pkg/server/serverpb/status.proto, line 1330 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwgyY9L0JoG9N6WNWlF:-MwgyY9L0JoG9N6WNWlG:b-l1szzv) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/serverpb/status.proto#L1330)):*
> ```protobuf
>   // Unix time range for aggregated statements.
>   int64 start = 6 [(gogoproto.nullable) = true];
>   int64 end = 7 [(gogoproto.nullable) = true];
> ```

nit: the field value should be `3` and `4` respectively here. 
___
*[pkg/server/serverpb/status.proto, line 1334 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwgyFpP9nWSeAv5QP3F:-MwgyFpP9nWSeAv5QP3G:b69csl1) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/serverpb/status.proto#L1334)):*
> ```protobuf
> 
> message StatementDetailsResponse {
>   message CollectedStatement {
> ```

(optional) nit: just as a personal preference, how about `CollectedStatementSummary` ? 


<!-- Sent from Reviewable.io -->

maryliag reviewed (commented):
It's similar to the overview statements, if the period selected has small amount of data it's fast, otherwise it might take several minutes

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-:-MwhApUD9IEw5yGSBW_v:bp9tnjx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng and @matthewtodd)

___
*[pkg/server/combined_statement_stats.go, line 351 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwgfauW5GFvmirKwZii:-MwhBDBk44MXwJTorCfr:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L351)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

nit: maybe move this function below `getStatementDetails` for reading flow?
</blockquote></details>

Done
___
*[pkg/server/combined_statement_stats.go, line 357 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mwgf2Ty-419F30fpqil:-MwhGXtM2Gdyj67zufbd:bvptkxf) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L357)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

Can you say more about the `len(args)+1` here and below? Is this code building a query "template" with embedded `$1` placeholders that will be filled in at execution time? If so, it's not clear to me why the same value is used throughout. EDIT: I see now that this function is also modifying `args` as it goes along, and that behavior surprised me. I hope there's a plainer way to write the code (I haven't taken a moment to think yet), but if not it's worth a comment highlighting the mechanism.
</blockquote></details>

Added a comment on the function description, explaining the format this is creating, also changed the order to add the arg first, so the length can used directly without the `+1`
___
*[pkg/server/combined_statement_stats.go, line 372 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mwgg-C7E38eITFwszig:-MwhGirq3dM0vXqqZCWg:b-hzhxzw) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L372)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

Does this mean we'll pass along both `(unset)` and `` (the empty string)? That feels suspicious to me.
</blockquote></details>

My idea here was that we treat `` as `unset` in the ui, so the ui might send `unset` if selected, and in this case we want to search for statements without app name, but in the very unlikely case that the actual name of the app is `unset` I decided to still keep it here. 
I decided to change and make that decision on the ui side now, so I'll be replacing by empty string before calling the endpoint
___
*[pkg/server/combined_statement_stats.go, line 376 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwggOny6joHdaszeR3a:-MwhHa0u2blKRAvBLMJk:b-2hxrje) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L376)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

Can this be done with a `strings.Join` or something? Except the modifying args in-place prevents that. Hmm.
</blockquote></details>

The reason for not using the join is exactly because of value on `$1` or `$2` and so on, as you mentioned.
___
*[pkg/server/combined_statement_stats.go, line 384 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mwh-B-c6GWXClFWPYOv:-MwhHoM8EN2iOz9NfCDM:b-wqvhnm) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L384)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

How about using the `IN` operator here. 
```
WHERE app_name IN (<val1>, <val2>, <val3>)
```
</blockquote></details>

That was actually my original solution, but during the replacement, instead of doing `app_name IN ('a', 'b', 'c')` it was creating `app_name IN (a, b, c)` (which would not return the correct value), even though it was specified as array of strings, I tried with different approaches of replacements and still didn't work, so I went with this options instead.
___
*[pkg/server/combined_statement_stats.go, line 459 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mwh03Ia6f_8jkRIaLNP:-MwhJQdn92h9JIHwZkR9:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L459)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

If we are only querying for one row, then we can use `QueryRowEx` here instead.
</blockquote></details>

Done
___
*[pkg/server/combined_statement_stats.go, line 479 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mwh-tf14Kle4e9qIFJQ:-MwhK4PtC1jgEg4diuKp:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L479)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

probably should be returning `serverError` here and below too as well
</blockquote></details>

Done
___
*[pkg/server/combined_statement_stats.go, line 525 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mwh1DdqCwyjAF1uANiP:-MwhKPaY3hBZb0GVIWmz:b8j14t3) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L525)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Simply assigning the `aggInterval` directly here
</blockquote></details>

`aggInterval` is `duration.Duration`, the value here requires Time.Duration, so this conversion is required
___
*[pkg/server/combined_statement_stats.go, line 527 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mwh0dwl3vvg1Lk3_FCB:-MwhLDb59zkGIBHhCMJK:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L527)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

nit: override the `metadata.Key.Query = queryPrettify` first, then you can simplify the assignment to be 

```golang

		statement = serverpb.StatementDetailsResponse_CollectedStatement{
			KeyData:             metadata.Key,
			AppNames:            appNames,
			AggregationInterval: time.Duration(aggInterval.Nanos()),
			Stats:               metadata.Stats,
		}

``` 


also, the `metadata` object here in fact also contains stats, might as well just call it `stats`. (I know we have been calling it `metadata` in other places of the code base, but that's really a mistake) 
</blockquote></details>

Done
___
*[pkg/server/combined_statement_stats.go, line 632 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mwh1yDW4Vlms1Ji8ExU:-MwhLXqO6s7crIlVcISK:bp52f0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/combined_statement_stats.go#L632)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

no need to use a slice here, since we know we only have 1 arg. 
</blockquote></details>

The QueryIteratorEx requires to be an array, so I would have to transform the planGist to an array anyway
___
*[pkg/server/serverpb/status.proto, line 1326 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mwgz1cV66ULWpiwb_25:-MwhM6wq5WktiK_SuxPa:b-neopld) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/serverpb/status.proto#L1326)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

nit: use the `gogoproto.customname` to override the Golang's field name to be `FingerprintID`, otherwise protoc will generate field name as `FingerprintId` 
</blockquote></details>

If the value is part of the query params , we do the custom name, but when is part of the path itself (you can see other examples on this same file: `DiagnosticsRequest`, `StoresRequest`, `RangeRequest`) we don't to because the gateway doesn't read customNames (there is also a comment on this in this file on the node_id usage for example)
___
*[pkg/server/serverpb/status.proto, line 1327 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwgzW5K-o-2yCCWRc3a:-MwhZR6L4zNO_-O9RIgN:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/serverpb/status.proto#L1327)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Just realized that this is not a single `app_name`, but a list of `app_name`s sent from the frontend. My would much prefer to have `repeated string app_names = 2;` here. 

String manipulation and the special logic to deal with the `(unset)` case is exclusively frontend concept. Leaking frontend concept into the backend code is only going to make the code harder to maintain going forward.
</blockquote></details>

Done
___
*[pkg/server/serverpb/status.proto, line 1330 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwgyY9L0JoG9N6WNWlF:-MwhMMYZ32qg3o2law-c:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/serverpb/status.proto#L1330)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

nit: the field value should be `3` and `4` respectively here. 
</blockquote></details>

Done
___
*[pkg/server/serverpb/status.proto, line 1334 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwgyFpP9nWSeAv5QP3F:-MwhM_tC1OFd90W3La-K:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/772108250c2624f59dde8472dd579cd876fe5f54/pkg/server/serverpb/status.proto#L1334)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

(optional) nit: just as a personal preference, how about `CollectedStatementSummary` ? 
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

matthewtodd reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (Maybe quickly confirm with @azhng on his `ExplainTreePlanNode` question?)

> Reviewed 2 of 5 files at r1, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-:-Mx-gEtQ9XfIQFr1mRe5:bdizlyi)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @Azhng and @maryliag)

___
*[-- commits, line 6 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mx-f6NX0S_AC3OYnsrh):*
nit: s/Aggregates/Aggregated/


<!-- Sent from Reviewable.io -->

Azhng reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> minus a few nits. Also do we need the release justification now? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-:-Mx0FyMy2KOB2h2beQEA:blovb7f)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @maryliag)

___
*[pkg/server/combined_statement_stats.go, line 480 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mx0FkNV9saOQk0StlRe:-Mx0FkNV9saOQk0StlRf:bq85w1c) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3f1ad32cedca9594ae540719a4ca172907e8711/pkg/server/combined_statement_stats.go#L480)):*
> ```Go
> 	if err = sqlstatsutil.DecodeStmtStatsMetadataJSON(metadataJSON, &statistics); err != nil {
> 		return statement, err
> 	}
> ```

nit: more `serverError` wrapping for the remaining of this function? 
___
*[pkg/server/serverpb/status.proto, line 1350 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwXJOWF8uAHq40mCzyf:-Mx0FU-34GT2RZRFRYmN:b-dw7nts) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56b9796795d8a3171cabb821140ca965698a102/pkg/server/serverpb/status.proto#L1350)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

That is correct for plans saved on our stats, where you select the plan and it's returned in one line as a json that you can decode to tree, but this one here comes form the `crdb_internal.decode_plan_gist()`, which returns a row per line, with them being strings, they're not rendered
</blockquote></details>

Ah I see. Might worth it call them `plan_gist` for unifying terminology? I'm also fine with it leaving as is.
___
*[pkg/server/serverpb/status.proto, line 1337 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mx0FJ6J7-2pLuNSGPEa:-Mx0FJ6J7-2pLuNSGPEb:b-p5btwz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3f1ad32cedca9594ae540719a4ca172907e8711/pkg/server/serverpb/status.proto#L1337)):*
> ```protobuf
>     cockroach.sql.StatementStatisticsKey key_data = 1 [(gogoproto.nullable) = false];
>     // Formatted query is the return of the key_data.query after prettify_statement.
>     // The value from the key_data cannot be replaced by the fomratted value, because is used as is for
> ```

nit: `s/forratted/formatted/` 


<!-- Sent from Reviewable.io -->

Azhng commented:
Also I know we are on a crunch right now, but we should improve the test coverage of these status endpoint during stability.

Azhng reviewed (commented):
Hmm also, is this endpoint omitted from `tenant_status.go` on purpose? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-:-Mx0H2AFBoY0mBiUosOW:b-q40tl8)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @maryliag)



<!-- Sent from Reviewable.io -->

maryliag reviewed (commented):
No need for release justification yet.
The tests are coming on the next PR and what do you mean it's omitted on tenant_status? It's here: https://github.com/cockroachdb/cockroach/pull/76592/files#diff-3e9c68308c7fc990de93cc07b0e2a7c2428f507efd59df9037e11027f83da6fdR551-R567

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-:-Mx0J4FH3f7oHSaKBpgS:b-gofpr3)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @matthewtodd)

___
*[pkg/server/combined_statement_stats.go, line 480 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-Mx0FkNV9saOQk0StlRe:-Mx0LsllAUqOAdIBwZUS:b-df857) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3f1ad32cedca9594ae540719a4ca172907e8711/pkg/server/combined_statement_stats.go#L480)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

nit: more `serverError` wrapping for the remaining of this function? 
</blockquote></details>

Done (for the rest of the file, not just this function)
___
*[pkg/server/serverpb/status.proto, line 1350 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-MwXJOWF8uAHq40mCzyf:-Mx0LhNR6mJMlJncUc_X:b-qrgoii) ([raw file](https://github.com/cockroachdb/cockroach/blob/e56b9796795d8a3171cabb821140ca965698a102/pkg/server/serverpb/status.proto#L1350)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Ah I see. Might worth it call them `plan_gist` for unifying terminology? I'm also fine with it leaving as is.
</blockquote></details>

It is an explain_plan, not a plan_gist. So the naming is using the right terminology, is just the format that is not JSON or ExplainTreePlanNode but a string instead.


<!-- Sent from Reviewable.io -->

maryliag reviewed (commented):
Actually, I added release justification

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-:-Mx0PJZeBDhl_bsVbJb0:bixqdf2)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @matthewtodd)



<!-- Sent from Reviewable.io -->

Azhng reviewed (commented):
Hmm, interesting. Somehow it's not being rendered on Reviewable :thinking: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-:-Mx0PThACl3QssXzGj3X:b-cnnhjc)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @matthewtodd)



<!-- Sent from Reviewable.io -->

Azhng reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76592#-:-Mx0PcmxDCcGhvgJ-8W4:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @matthewtodd)



<!-- Sent from Reviewable.io -->

maryliag commented:
bors r+

ajwerner commented:
Seems like this and https://github.com/cockroachdb/cockroach/pull/74377 skew with each-other in terms of generated code.

maryliag commented:
I can see you cancelled the other PR, thank for looking into it
